WebGPU API docs (#24202)
mdn/content: files/en-us/web/api/gpuadapter/requestdevice/index.md

---
title: GPUAdapter.requestDevice()
slug: Web/API/GPUAdapter/requestDevice
page-type: web-api-instance-method
status:
  - experimental
browser-compat: api.GPUAdapter.requestDevice
---

{{APIRef("WebGPU API")}}{{SeeCompatTable}}

The **`requestDevice()`** method of the

> {{domxref("GPUAdapter")}} interface returns a {{jsxref("Promise")}} that fulfills with a {{domxref("GPUDevice")}} object via which you can start using the functionality of a device GPU.

Reviewer: ```suggestion
{{domxref("GPUAdapter")}} interface returns a {{jsxref("Promise")}} that fulfills with a {{domxref("GPUDevice")}} object, which is the primary interface for communicating with the GPU.
```

Author: updated

## Syntax

```js-nolint
requestDevice()
requestDevice(descriptor)
```

### Parameters

- `descriptor` {{optional_inline}}
  - : An object containing the following properties:
    - `defaultQueue` {{optional_inline}}
      - : An object that provides information for the device's default {{domxref("GPUQueue")}} (as returned by {{domxref("GPUDevice.queue")}}). This object has a single property — `label` — which provides the default queue with a {{domxref("GPUQueue.label", "label")}} value. If no value is provided, this defaults to an empty object, and the default queue's label will be an empty string.
    - `label` {{optional_inline}}

>       - : A string providing an identifying label that can be used to identify the object, for example in {{domxref("GPUError")}} messages or console warnings.

Reviewer: "identifying label that can be used to identify" is redundant. Should be OK to drop one of them.

```suggestion
      - : A string providing a label that can be used to identify the object, for example in {{domxref("GPUError")}} messages or console warnings.
```

Author: Yup, good call. I've updated it to use your text.

  }

  const adapter = await navigator.gpu.requestAdapter();
  if (!adapter) {
    throw Error('Couldn\'t request WebGPU adapter.');
  }

  const device = await adapter.requestDevice();

  // ...

}
```

### Requesting specific features and limits

In the following code we:

1. Check whether a {{domxref("GPUAdapter")}} has the `texture-compression-astc` feature available. If so, we push it into the array of `requiredFeatures`.
2. Query the {{domxref("GPUAdapter.limits")}} values of `maxTextureDimension3D` and `maxBindGroups` and add those maximum limits to the `requiredLimits` object.
3. Request a device with those feature and limit requirements, plus a `defaultQueue` object.

```js
async function init() {
  if (!navigator.gpu) {
    throw Error('WebGPU not supported.');
  }

  const adapter = await navigator.gpu.requestAdapter();
  if (!adapter) {
    throw Error('Couldn\'t request WebGPU adapter.');
  }

  const requiredFeatures = [];

  if (adapter.features.has('texture-compression-astc')) {
    requiredFeatures.push('texture-compression-astc')
  }

  const requiredLimits = {};

  requiredLimits.maxTextureDimension3D = adapter.limits.maxTextureDimension3D;

>   requiredLimits.maxBindGroups = adapter.limits.maxBindGroups;

Reviewer: If you make the suggested change to the limits example above the same thing would apply here.

Author: Yup, done.
